# Finding a Place for a Restaurant in Raleigh

Introduction/Business Problem:
In the city of Raleigh, if one wants to find an area to create a restaurant venue, where would you recommend opening one?  There are a lot of restaurants in Raleigh, especially since it is the home of one of the state’s largest universities, NC State. In addition, there are 15 other colleges and universities within local area. The background of the problem is, where might someone open a profitable restaurant venue that would maximize profits, as you want to not only take advantage of the fact that there are a significant amount of young people in the area, but so that you do not lose profits or customers to other surrounding entertainment venues?

The basic idea is to find a suitable location for an restaurant that is within Raleigh, that is suitable for not only business, but takes advantage of the proximity to colleges and universities in the area.

Data Collection:
The project will collect data from the following sources:

•	Foursquare API: use Foursquare to collect data from Raleigh and surrounding areas within 50 miles 
•	Raleigh Open Data: contains API for both Colleges and Universities and Restaurants

Raleigh:
Raleigh geographical area covers 144.8 square miles. The objective is to find probable areas that may be advantageous to opening a restaurant.

To begin, we need to identify the tools that are needed to pull data from the file.

In [6]:
# import the libraries we need
import pandas as pd
import geopy
from geopy.geocoders import Nominatim
import numpy as np
!pip install folium
import folium

In [76]:
import pandas as pd
import requests
from bs4 import BeautifulSoup
from tabulate import tabulate

file_name='https://opendata.arcgis.com/datasets/cfabd5fb515b4b45a2714d48fbcd2dd5_0.csv?outSR=%7B%22latestWkid%22%3A2264%2C%22wkid%22%3A102719%7D&session=1196661311.1585645749.csv'
df=pd.read_csv(file_name)


In [77]:
print('shape: ',df.shape)
print('columns: ', df.columns)
df.head()

shape:  (18, 6)
columns:  Index(['X', 'Y', 'OBJECTID', 'ADDRESS', 'CITY', 'COLLEGENAME'], dtype='object')


X              Y  OBJECTID                   ADDRESS     CITY  \
0  2.097579e+06  741713.299902         1  2101 Hillsborough Street  Raleigh   
1  2.092281e+06  744629.599844         2  3800 Hillsborough Street  Raleigh   
2  2.095116e+06  736981.100049         3      1007 Capability Road  Raleigh   
3  2.107643e+06  735692.700066         4          118 South Street  Raleigh   
4  2.112693e+06  741014.500118         5       1315 Oakwood Avenue  Raleigh   

                         COLLEGENAME  
0  NC State University - Main Campus  
1                   Meredith College  
2             NCSU Centennial Campus  
3                    Shaw University  
4              St. Augustine College

In [37]:
df.shape

(3481, 15)

# Methodology

For the below, wanted to provide location data for all the colleges and universities located in Raleigh.

In [80]:
df3 = df[['COLLEGENAME', 'X', 'Y']]
df3.head(20)

COLLEGENAME             X  \
0                   NC State University - Main Campus  2.097579e+06   
1                                    Meredith College  2.092281e+06   
2                              NCSU Centennial Campus  2.095116e+06   
3                                     Shaw University  2.107643e+06   
4                               St. Augustine College  2.112693e+06   
5                       Southeastern Baptist Seminary  2.144425e+06   
6         Wake Technical College Southern Wake Campus  2.087623e+06   
7                            William Peace University  2.107440e+06   
8         Wake Technical College Northern Wake Campus  2.134938e+06   
9          Wake Technical College Western Wake Campus  2.060739e+06   
10  Wake Technical College Perry Health Sciences C...  2.122848e+06   
11                  Wake Technical College RTP Campus  2.049020e+06   
12  Wake Technical College Public Safety Education...  2.104581e+06   
13           Vernon Malone College and Career Academy  2.105868e+06   
14                           Campbell University ext.  2.055586e+06   
15                     Campbell University Law School  2.105852e+06   
16   Wake Technical College Beltline Education Center  2.114660e+06   
17  Wake Technical College Eastern Wake Education ...  2.205384e+06   

                Y  
0   741713.299902  
1   744629.599844  
2   736981.100049  
3   735692.700066  
4   741014.500118  
5   812169.599934  
6   692108.900095  
7   742210.100089  
8   770863.000133  
9   711927.599959  
10  740698.100161  
11  766080.999854  
12  722558.999956  
13  729089.000019  
14  758889.400064  
15  739116.799846  
16  755005.982634  
17  757267.557179

Next, we need to pull the data for restaurants in the Raleigh Area.

In [96]:
import pandas as pd
import requests
from bs4 import BeautifulSoup
from tabulate import tabulate

file_name='https://opendata.arcgis.com/datasets/124c2187da8c41c59bde04fa67eb2872_0.csv?session=1285114555.1585645226.csv'
df=pd.read_csv(file_name)


In [97]:
print('shape: ',df.shape)
print('columns: ', df.columns)
df.head()

shape:  (3481, 15)
columns:  Index(['OBJECTID', 'HSISID', 'NAME', 'ADDRESS1', 'ADDRESS2', 'CITY', 'STATE',
       'POSTALCODE', 'PHONENUMBER', 'RESTAURANTOPENDATE', 'FACILITYTYPE',
       'PERMITID', 'X', 'Y', 'GEOCODESTATUS'],
      dtype='object')


OBJECTID      HSISID                                NAME  \
0    509616  4092018236                    Tacos Don Marcos   
1    509841  4092050026  St Mary Magdalene School Cafeteria   
2    509842  4092015002                BAAN THAI RESTAURANT   
3    509843  4092014554                        KFC #J120103   
4    509844  4092014694                  Brueggers's Bagels   

               ADDRESS1 ADDRESS2    CITY STATE POSTALCODE  \
0  9629 BITTER MELON DR      NaN  ANGIER    NC      27501   
1       625 Magdala PL       NaN    APEX    NC      27502   
2      758 WILLIAMS ST       NaN    APEX    NC      27502   
3    1403 W Williams ST      NaN    APEX    NC      27502   
4    760 W Williams ST       NaN    APEX    NC      27502   

               PHONENUMBER        RESTAURANTOPENDATE  \
0           (919) 272-7062  2019-11-19T00:00:00.000Z   
1  (919) 657-4800 ext. 298  2004-08-13T00:00:00.000Z   
2           (919) 629-6399  2007-11-09T00:00:00.000Z   
3           (919) 303-7997  2006-04-25T00:00:00.000Z   
4           (919) 367-7720  2006-09-29T00:00:00.000Z   

                FACILITYTYPE  PERMITID          X          Y GEOCODESTATUS  
0                 Restaurant     18813 -78.747986  35.539550             M  
1  Private School Lunchrooms      5112 -78.874487  35.724227             M  
2                 Restaurant      1953 -78.864641  35.738078             M  
3                 Restaurant      8630 -78.877270  35.750153             M  
4                 Restaurant     19940 -78.862761  35.738032             M

In [98]:
df.shape

(3481, 15)

In [99]:
df4= df[['NAME', 'X', 'Y']]
df4.head()

NAME          X          Y
0                    Tacos Don Marcos -78.747986  35.539550
1  St Mary Magdalene School Cafeteria -78.874487  35.724227
2                BAAN THAI RESTAURANT -78.864641  35.738078
3                        KFC #J120103 -78.877270  35.750153
4                  Brueggers's Bagels -78.862761  35.738032

We also need to identify and map the city and gain the coordinates.

In [39]:
address = 'Raleigh City'
geolocator = Nominatim(user_agent='to_explorer')
location = geolocator.geocode(address)
print(location.latitude, location.longitude)

35.7803977 -78.6390989


In [41]:
CHD_LATITUDE = '35.7803977'
CHD_LONGITUDE = '-78.6390989'
print('The geograpical coordinates of Raleigh are {}, {}.'.format(CHD_LATITUDE, CHD_LONGITUDE))

The geograpical coordinates of Raleigh are 35.7803977, -78.6390989.


In [55]:
raleigh_map = folium.Map(location = [CHD_LATITUDE, CHD_LONGITUDE], zoom_start = 13)
folium.Marker([CHD_LATITUDE, CHD_LONGITUDE]).add_to(raleigh_map)

In [57]:
raleigh_map

After identifying Raleigh on the map, we then need to identify the venues using Foursquare as well as map the venues.

In [59]:
foursquare_client_ID = 'VAYCGD5DPRVQT5URMRBXLYILFFXVGOAVYNAMPXYGABJF1MV0'
foursquare_client_secret = 'I3TRZNORMFIVXTTOIJEFCBQC0SYS2ED4UQDMEV1UYWZLXYHS'
radius = 80467 # 50 mi = 80.4672Km
no_of_venues = 150
version = '20200331' # Current date

In [60]:
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [61]:
import numpy as np
import pandas as pd

import matplotlib.pyplot as plt
import matplotlib.cm as cm
import matplotlib.colors as colors

from pandas.io.json import json_normalize
import requests

pd.set_option('display.max_rows', None)

offset = 0
total_venues = 0
foursquare_venues = pd.DataFrame(columns = ['name', 'categories', 'lat', 'lng'])

while (True):
    url = ('https://api.foursquare.com/v2/venues/explore?client_id={}'
           '&client_secret={}&v={}&ll={},{}&radius={}&limit={}&offset={}').format(foursquare_client_ID, 
                                                                        foursquare_client_secret, 
                                                                        version, 
                                                                        CHD_LATITUDE, 
                                                                        CHD_LONGITUDE, 
                                                                        radius,
                                                                        no_of_venues,
                                                                        offset)
    result = requests.get(url).json()
    venues_fetched = len(result['response']['groups'][0]['items'])
    total_venues = total_venues + venues_fetched
    print("Total {} venues fetched within a total radius of {} Km".format(venues_fetched, RADIUS/1000))

    venues = result['response']['groups'][0]['items']
    venues = json_normalize(venues)

    # Filter the columns
    filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
    venues = venues.loc[:, filtered_columns]

    # Filter the category for each row
    venues['venue.categories'] = venues.apply(get_category_type, axis = 1)

    # Clean all column names
    venues.columns = [col.split(".")[-1] for col in venues.columns]
    foursquare_venues = pd.concat([foursquare_venues, venues], axis = 0, sort = False)
    
    if (venues_fetched < 100):
        break
    else:
        offset = offset + 100

foursquare_venues = foursquare_venues.reset_index(drop = True)
print("\nTotal {} venues fetched".format(total_venues))

Total 100 venues fetched within a total radius of 80.467 Km
Total 100 venues fetched within a total radius of 80.467 Km
Total 46 venues fetched within a total radius of 80.467 Km

Total 246 venues fetched


In [62]:
raleigh_map = folium.Map(location = [CHD_LATITUDE, CHD_LONGITUDE], zoom_start = 13)

for name, latitude, longitude in zip(foursquare_venues['name'], foursquare_venues['lat'], foursquare_venues['lng']):
    label = '{}'.format(name)
    label = folium.Popup(label, parse_html = True)
    folium.CircleMarker(
        [latitude, longitude],
        radius = 5,
        popup = label,
        color = 'green',
        fill = True,
        fill_color = '#3186cc',
        fill_opacity = 0.7,
        parse_html = False).add_to(raleigh_map)  


raleigh_map

In [146]:
import numpy as np
import pandas as pd

import matplotlib.pyplot as plt
import matplotlib.cm as cm
import matplotlib.colors as colors

from pandas.io.json import json_normalize
import requests

pd.set_option('display.max_rows', None)

offset = 0
total_venues = 0
foursquare_venues = pd.DataFrame(columns = ['name', 'categories', 'lat', 'lng'])

while (True):
    url = ('https://api.foursquare.com/v2/venues/explore?client_id={}'
           '&client_secret={}&v={}&ll={},{}&radius={}&limit={}&offset={}').format(foursquare_client_ID, 
                                                                        foursquare_client_secret, 
                                                                        version, 
                                                                        CHD_LATITUDE, 
                                                                        CHD_LONGITUDE, 
                                                                        radius,
                                                                        no_of_venues,
                                                                        offset)
    result = requests.get(url).json()
    venues_fetched = len(result['response']['groups'][0]['items'])
    total_venues = total_venues + venues_fetched
    print("Total {} venues fetched within a total radius of {} Km".format(venues_fetched, RADIUS/1000))

    venues = result['response']['groups'][0]['items']
    venues = json_normalize(venues)

    # Filter the columns
    filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
    venues = venues.loc[:, filtered_columns]

    # Filter the category for each row
    venues['venue.categories'] = venues.apply(get_category_type, axis = 1)

    # Clean all column names
    venues.columns = [col.split(".")[-1] for col in venues.columns]
    foursquare_venues = pd.concat([foursquare_venues, venues], axis = 0, sort = False)
    
    if (venues_fetched < 100):
        break
    else:
        offset = offset + 100

foursquare_venues = foursquare_venues.reset_index(drop = True)
print("\nTotal {} venues fetched".format(total_venues))

Total 100 venues fetched within a total radius of 80.467 Km
Total 100 venues fetched within a total radius of 80.467 Km
Total 46 venues fetched within a total radius of 80.467 Km

Total 246 venues fetched


We need to identify the proximity of the universities to the restaurants. However, there are 17 colleges and universities, and 3481 restaurants in the Raleigh area. There are 246 venues within a 50 mile radius of Raleigh. So we now need to reduce the amount of data that we will use, as it goes beyond the scope of what the Capstone intends. The data isn't as compatible as I believed when I began. I now will adjust the scope of the project to the scope of my experience. I will use a specific location, NC State University, and will attempt to determine the venues surrounding it.

In [154]:
!pip install geopandas
!pip install geopy

     |████████████████████████████████| 931kB 14.8MB/s eta 0:00:01
     |████████████████████████████████| 10.4MB 47.5MB/s eta 0:00:01
     |████████████████████████████████| 1.8MB 45.0MB/s eta 0:00:01
     |████████████████████████████████| 14.7MB 25.9MB/s eta 0:00:01


In [155]:
file_name='https://opendata.arcgis.com/datasets/cfabd5fb515b4b45a2714d48fbcd2dd5_0.csv?outSR=%7B%22latestWkid%22%3A2264%2C%22wkid%22%3A102719%7D&session=1196661311.1585645749.csv'
df=pd.read_csv(file_name)

In [157]:
print('shape: ',df.shape)
print('columns: ', df.columns)
df.head()

shape:  (18, 6)
columns:  Index(['X', 'Y', 'OBJECTID', 'ADDRESS', 'CITY', 'COLLEGENAME'], dtype='object')


X              Y  OBJECTID                   ADDRESS     CITY  \
0  2.097579e+06  741713.299902         1  2101 Hillsborough Street  Raleigh   
1  2.092281e+06  744629.599844         2  3800 Hillsborough Street  Raleigh   
2  2.095116e+06  736981.100049         3      1007 Capability Road  Raleigh   
3  2.107643e+06  735692.700066         4          118 South Street  Raleigh   
4  2.112693e+06  741014.500118         5       1315 Oakwood Avenue  Raleigh   

                         COLLEGENAME  
0  NC State University - Main Campus  
1                   Meredith College  
2             NCSU Centennial Campus  
3                    Shaw University  
4              St. Augustine College

In [160]:
from geopy.geocoders import Nominatim
geolocator = Nominatim()
location = geolocator.geocode("2101 Hillsborough Street NC")
print(location.address)


/opt/conda/envs/Python36/lib/python3.6/site-packages/ipykernel/__main__.py:2: DeprecationWarning: Using Nominatim with the default "geopy/1.18.1" `user_agent` is strongly discouraged, as it violates Nominatim's ToS https://operations.osmfoundation.org/policies/nominatim/ and may possibly cause 403 and 429 HTTP errors. Please specify a custom `user_agent` with `Nominatim(user_agent="my-application")` or by overriding the default `user_agent`: `geopy.geocoders.options.default_user_agent = "my-application"`. In geopy 2.0 this will become an exception.
  from ipykernel import kernelapp as app


2101, Hillsborough Street, Oberlin, Wake County, North Carolina, 27607, United States of America


In [161]:
print((location.latitude, location.longitude))

(35.786319, -78.663314)


In [164]:
foursquare_client_ID = 'VAYCGD5DPRVQT5URMRBXLYILFFXVGOAVYNAMPXYGABJF1MV0'
foursquare_client_secret = 'I3TRZNORMFIVXTTOIJEFCBQC0SYS2ED4UQDMEV1UYWZLXYHS'
radius = 80467 # 50 mi = 80.4672Km
no_of_venues = 10
version = '20200331' # Current date

In [165]:
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [274]:
import numpy as np
import pandas as pd

import matplotlib.pyplot as plt
import matplotlib.cm as cm
import matplotlib.colors as colors

from pandas.io.json import json_normalize
import requests

pd.set_option('display.max_rows', None)

offset = 0
total_venues = 0
foursquare_venues = pd.DataFrame(columns = ['name', 'categories', 'lat', 'lng'])

while (True):
    url = ('https://api.foursquare.com/v2/venues/explore?client_id={}'
           '&client_secret={}&v={}&ll={},{}&radius={}&limit={}&offset={}').format(foursquare_client_ID, 
                                                                        foursquare_client_secret, 
                                                                        version, 
                                                                        CHD_LATITUDE, 
                                                                        CHD_LONGITUDE, 
                                                                        radius,
                                                                        no_of_venues,
                                                                        offset)
    result = requests.get(url).json()
    venues_fetched = len(result['response']['groups'][0]['items'])
    total_venues = total_venues + venues_fetched
    print("Total {} venues fetched within a total radius of {} Km".format(venues_fetched, RADIUS/1000))

    venues = result['response']['groups'][0]['items']
    venues = json_normalize(venues)

    # Filter the columns
    filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
    venues = venues.loc[:, filtered_columns]

    # Filter the category for each row
    venues['venue.categories'] = venues.apply(get_category_type, axis = 1)

    # Clean all column names
    venues.columns = [col.split(".")[-1] for col in venues.columns]
    foursquare_venues = pd.concat([foursquare_venues, venues], axis = 0, sort = False)
    
    if (venues_fetched < 100):
        break
    else:
        offset = offset + 100

foursquare_venues = foursquare_venues.reset_index(drop = True)
print("\nTotal {} venues fetched".format(total_venues))

Total 10 venues fetched within a total radius of 80.467 Km

Total 10 venues fetched


In [275]:
df.shape

(18, 6)

In [276]:
df.groupby('ADDRESS').size()

ADDRESS
1007 Capability Road            1
10908 Chapel Hill Rd            1
118 South Street                1
1315 Oakwood Avenue             1
15 E. Peace Street              1
2101 Hillsborough Street        1
219 S. Wingate Street           1
2200 South Wilmington Street    1
225 Hillsborough St             1
2901 Holston Ln                 1
3200 Bush St                    1
321 Chapanoke Rd                1
3434 Kildaire Farm Rd           1
3800 Hillsborough Street        1
519 N Industrial Dr             1
6600 Louisburg Rd               1
808 Aviation Pkwy #1100         1
9101 Fayetteville Road          1
dtype: int64

In [277]:
df2 = df[['ADDRESS', 'X', 'Y']]
df2.head()

ADDRESS             X              Y
0  2101 Hillsborough Street  2.097579e+06  741713.299902
1  3800 Hillsborough Street  2.092281e+06  744629.599844
2      1007 Capability Road  2.095116e+06  736981.100049
3          118 South Street  2.107643e+06  735692.700066
4       1315 Oakwood Avenue  2.112693e+06  741014.500118

In [278]:
df2.columns = ['id', 'Latitude', 'Longitude']
df2.index = range(len(df2))
df2.head()

id      Latitude      Longitude
0  2101 Hillsborough Street  2.097579e+06  741713.299902
1  3800 Hillsborough Street  2.092281e+06  744629.599844
2      1007 Capability Road  2.095116e+06  736981.100049
3          118 South Street  2.107643e+06  735692.700066
4       1315 Oakwood Avenue  2.112693e+06  741014.500118

In [279]:
df3 = df2.loc[:99, :]
df3.shape

(18, 3)

In [280]:
def getNearbyVenues(id, latitudes, longitudes, radius = 80467):
    
    venues_list = []
    for index, lat, lng in zip(id, latitudes, longitudes):
        print(index)
    
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    latitude, 
    longitude, 
    radius, 
    LIMIT)
results = requests.get(url).json()
nearby_venues = pd.DataFrame()

venues = results['response']['groups'][0]['items']
nearby_venues = json_normalize(venues)

In [281]:
# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]
# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)
# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]
nearby_venues.head()

name                       categories  \
0  North Carolina Museum of Natural Sciences                   Science Museum   
1      Bida Manda Laotian Restaurant and Bar                 Asian Restaurant   
2                Artspace Visual Arts Center                      Art Gallery   
3           North Carolina Museum of History                   History Museum   
4                  Beasley's Chicken + Honey  Southern / Soul Food Restaurant   

         lat        lng  
0  35.782217 -78.639375  
1  35.777227 -78.636812  
2  35.775576 -78.636335  
3  35.781516 -78.638747  
4  35.776968 -78.638175

In [282]:
nearby_venues.tail()

name      categories        lat        lng
5         Burial Beer Co.         Brewery  35.775241 -78.632247
6  Nature Research Center  Science Museum  35.782486 -78.640745
7       Logan Trading Co.   Garden Center  35.789426 -78.640663
8     Marbles Kids Museum          Museum  35.778624 -78.636033
9            lucettegrace          Bakery  35.777266 -78.640189

In [295]:
import numpy as np
import pandas as pd

import matplotlib.pyplot as plt
import matplotlib.cm as cm
import matplotlib.colors as colors

from pandas.io.json import json_normalize
import requests
venues_map = folium.Map(location=[35.786319, -78.663314], zoom_start=15)

folium.CircleMarker(
    [latitude, longitude],
    radius=5,
    color='red',
    popup='Raleigh',
    fill = True,
    fill_color = 'red',
    fill_opacity = 0.6
).add_to(venues_map)
# add all venues as blue circle markers
for lat, lng, label in zip(nearby_venues.lat, nearby_venues.lng, nearby_venues.categories):
    folium.CircleMarker(
        [lat, lng],
        radius=10,
        color='blue',
        popup=label,
        fill = True,
        fill_color='blue',
        fill_opacity=0.6
    ).add_to(venues_map)

In [296]:
venues_map

# Results
Based on the data collected, a probable area to set up a restaurant near NC State in Raleigh would be within the range of the warehouse district.

Conclusion:
Discussion: Trial, Error, and Changing Scope:

I now understand that while I have the training to apply certain tools, it takes experience to apply them effectively. The original scope was to find the best location for a restaurant using all data from surrounding colleges and universities. However, after much confusion and frustration, I realized that I do not have the experience to work with complex tables that do not match as well as hoped.  From researching data and results, it would have been better to select data that was compatible with each other, i.e. both having longitude and latitude data on it. From the data, we can see a number of surrounding sources, but not all of them are restaurants.